File: 5.prefix-sum/연경/3.subarray-sum-equals-k.py
## https://leetcode.com/problems/subarray-sum-equals-k/
from typing import List
from collections import defaultdict
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        answer = 0
        count = defaultdict(int)

        p = 0
        for num in nums:
            p += num
            if p == k:
                answer += 1

            if p-k in count:
                answer += count[p-k]
            count[p] += 1
        return answer


if __name__ == "__main__":
    nums = [1, 2, 3]
    k = 3
    print(Solution().subarraySum(nums, k))

File: 5.prefix-sum/동규/3.subarray-sum-equals-k.js
// https://leetcode.com/problems/subarray-sum-equals-k/

var subarraySum = function (nums, k) {
  let count = 0,
    sum = 0;
  const map = new Map();
  map.set(0, 1);
  for (let i = 0; i < nums.length; i++) {
    sum += nums[i];
    if (map.has(sum - k)) count += map.get(sum - k);
    map.set(sum, (map.get(sum) || 0) + 1);
  }
  return count;
};
